[cmd2] Don't mutate scheduled commands set during iteration

kytpbs opened the pull request:
Java counterpart to #7554.

Found together with @PeterJohnson.

kytpbs committed:
fix scheduledCommands.erase() being called in the same sets iterator

Starlight220 reviewed (commented):
Wasn't the iteration being done via an iterator (and then the removal also being done through the iterator)?

There's already `endingCommands`, and this adds another `toRemove`; I feel we're hacking this implementation patch on patch and it's not great.

This diverges from Java.

This changes behavior, as with this `isScheduled` will return `true` until the following iteration and not immediately when `isFinished` returns false. As an example of practical change, proxies will now end the iteration after the proxied command (which tbh might be better than the current situation where it depends on the map iteration order).

Add unit tests based on the code used to find this.

PeterJohnson commented:
> Wasn't the iteration being done via an iterator (and then the removal also being done through the iterator)?

Yes, but this doesn't work with SmallSet--it invalidates the iterator, which means the for loop next iteration is incrementing from an invalid iterator (UB).

Starlight220 commented:
`SmallSet` doesn't have an `Iterator.remove` link Java?

PeterJohnson commented:
No.  Neither does std::set.

kytpbs committed:
added comment explaning why we defer from java

this is also to retrigger workflows

kytpbs commented:
/format

github-actions[bot] committed:
Formatting fixes

KangarooKoala reviewed (commented):
Good catch!

However, considering that this certainly does have a change in behavior, Java should be changed to match. (We've had enough issues with differences between Java and C++, so it's best to keep them in sync as much as possible.) I'll note that [robotpy's command scheduler](https://github.com/robotpy/robotpy-commands-v2/blob/main/commands2/commandscheduler.py) avoids this issue by iterating over a (shallow) copy of the map/dictionary, but it still has the scheduling and canceling queues (presumably for compatibility?).

Additionally, I'm nervous about partially delaying ending the command- It feels like at least one person will run into an edge case where a "scheduled" command not being linked to any requirements causes strange and difficult-to-debug behavior.

PeterJohnson commented on wpilibNewCommands/src/main/native/cpp/frc2/command/CommandScheduler.cpp:
```diff
@@ -196,6 +196,7 @@ void CommandScheduler::Run() {
   m_impl->inRunLoop = true;
   bool isDisabled = frc::RobotState::IsDisabled();
   // Run scheduled commands, remove finished commands.
+  auto toRemove = std::vector<Command*>(m_impl->scheduledCommands.size());
```
How often are commands removed?  I would recommend using a SmallVector here with say 8 elements, and not reserving at all.  That will avoid allocating on most calls.

kytpbs replied:
> The probable intent is to guarantee the vector has enough space to store all of the commands
That's exactly what I was trying to do.  Thank you for noticing.


> I would recommend using a SmallVector here with say 8 elements

I mean even with a vector, the maximum we'd likely do is allocate at startup, and subsequent reserve calls would invoke malloc, but this wouldn't trigger a system call since we'd already have the space.
and if we were to switch this to Small::Vector

should we do it like this?

```suggestion
  wpi::SmallVector<Command*, 10> toRemove;
```
or like this this?
```suggestion
  wpi::SmallVector<Command*, 10> toRemove;
  toRemove.reserve(m_impl->scheduledCommands.size());
```
(the 10 can be changed to whatever seems appropriate)

KangarooKoala replied:
Default-constructing a `std::vector` and then calling `reserve()` will almost certainly allocate. (I'm not sure what compilers to test, but from https://godbolt.org/z/9sbevjT64, it seems that on x86-64 clang 14.0.0 at least, the starting capacity is 0, so reserving more space will need to allocate more memory)

In contrast, constructing a `SmallVector` and not reserving space (no `reserve()` call) will not allocate any heap memory (instead using stack memory, which is extremely cheap to get) unless the number of commands actually being ended exceeds the template parameter (which is highly unlikely- a typical maximum number of subsystems is 5 to 7, and it's exceedingly rare for all of them to end simultaneously with additional commands also ending simultaneously), in which case it will almost certainly only allocate once. (The number of elements allocated doubles each time the vector grows, so 20 (scheduler-managed) commands would have to end in the same call to `run()` to get just 2 allocations.)

kytpbs replied:
but at the same time if a team has decided to go with C++ I am guessing they are going for the highest performance they can get (_as that is mostly the biggest pro of going with C++ in the first place_). which might mean due to our implementation if they have a lot of commands scheduled, it might cause slowdowns that may make the team's devs question why. 

We may also go with robot.py's technique of copying the set and iterating over that instead, which would make erase functions a non-issue but will have the problem of copying a `smallSet` every single periodic loop. Still, if the small set is only stack allocated then it should be fine.

KangarooKoala replied:
I'm not sure I communicated my point clearly enough- What I'm saying is that using a `SmallVector` without reserving will be strictly faster than using a `std::vector` in the vast majority of use cases and at least as fast in the rest, unless someone is misusing the library (in which case we can only help so much).

If you think that more than 16 commands ending all in the same loop is something that we should support, then that's a different question, but my understanding of what you wrote (which might be wrong) is that you were concerned about performance overall without addressing the likelihood of more than 16 commands ending in the same loop. (The 16 comes from 2*8, assuming we use `SmallVector<Command*, 8>`. If you use `SmallVector<Command*, 10>` then the limit becomes 20 commands ending simultaneously before using a small vector gets slower.)

kytpbs committed:
use small Vector instead

size is set to 10, as I don't think any normal team is going to end 20 commands at the same time...

kytpbs commented:
The way [robot.py](https://github.com/robotpy/robotpy-commands-v2/blob/main/commands2/commandscheduler.py) did it seems like a better way to mitigate this issue, and `SmallSet<Command*>(scheduledCommands)` should create a copy of the smallSet, and I don't think the smallSet will be big enough that this will have a performance impact.

What do you guys think?

KangarooKoala commented:
I agree that copying `scheduledCommands` would be the cleaner solution. Some things to remember for whoever implements that:
* Make the same change to Java (I think iterating over `m_scheduledCommands.toArray(new Command[0])` would be the most performant way to copy the set)
* Add a check to the start of the run loop to skip commands that are no longer scheduled
* Remove the `inRunLoop`, `toSchedule`, `toCancelCommands`, and `toCancelInterruptors` member variables

TheTripleV commented:
Whatever behavior is being chosen, if it's going to being a part of the "spec", there should be tests that fail currently and pass afterwards added.

kytpbs committed:
create a new SmallSet instead

this is similar to the way robot.py has done it, and because we are only copying a pointer array, it shouldn't have a big impact on performance.

kytpbs committed:
create a copy of the `m_scheduledCommands`

this is made to be consistent with C++ due to it not having iterator.Remove() or similar

kytpbs committed:
remove concurrent avoiding modification member variables (C++)

kytpbs committed:
remove concurrent avoiding modification member variables (Java)

kytpbs commented:
/format

github-actions[bot] committed:
Formatting fixes

KangarooKoala commented on wpilibNewCommands/src/main/java/edu/wpi/first/wpilibj2/command/CommandScheduler.java:
```diff
@@ -274,11 +262,14 @@ public void run() {
     boolean isDisabled = RobotState.isDisabled();
     // Run scheduled commands, remove finished commands.
-    for (Iterator<Command> iterator = m_scheduledCommands.iterator(); iterator.hasNext(); ) {
+    for (Iterator<Command> iterator = Set.copyOf(m_scheduledCommands).iterator();
+        iterator.hasNext(); ) {
       Command command = iterator.next();
```
```suggestion
    for (Command command : Set.copyOf(m_scheduledCommands)) {
```
Since we don't remove via the iterator, we don't need to keep track of it separately.

kytpbs replied:
wasn't sure about deleting it but makes sense to me too so done!

KangarooKoala reviewed (approved):
Looks good to me! (For what it's worth)
I'll just note down the behavioral differences for posterity, but I think these are acceptable (even positive) changes:
* `command.isScheduled()` is `false` when `command.end(boolean)` and the finish or interrupt callbacks are called.
* When scheduling a command from the run loop, `command.initialize()` and the schedule callbacks are called immediately, not after the run loop. It remains the case that `command.execute()` will only be called from the next call to `scheduler.run()`.
* When cancelling a command from the run loop , `command.end(true)` and the finish or interrupt callbacks are called immediately, not after the run loop (and the queued scheduled commands are processed). Notably, if a command (call it A) cancels another command (call it B) that was scheduled after A (and therefore will be processed after A), B is **not** be processed (`execute()` called and `isFinished()` checked) in the same run loop.

kytpbs committed:
remove iterator as we don't need it

Co-authored-by: Joseph Eng <91924258+KangarooKoala@users.noreply.github.com>

kytpbs commented:
/format

github-actions[bot] committed:
Formatting fixes

PeterJohnson commented:
Set.copyOf is memory allocation heavy; this is what the implementation is:
```java
return (Set<E>)Set.of(new HashSet<>(coll).toArray());
```
So it first creates a HashSet, then creates an array of its elements, then creates a Set from that array.

For how we're using it here, toArray() would be more efficient.

KangarooKoala commented:
Note that `toArray(T[])` will automatically try to reuse the passed array, so a simple way to reduce allocations would be to update the array with  `m_composedCommandsCopy = m_composedCommands.toArray(m_composedCommandsCopy)`. (There's some options we could explore such as the initial array size and growing the array by more than is initially required, but I don't know if it would necessarily be worth it)

Starlight220 reviewed (changes requested):
This needs so many tests.
`SchedulingRecursionTest` passing is a good sign (or a very bad one, if the tests there aren't working).

All behavior changes, edge cases, and so on need to be tested. Rigorously.

kytpbs commented:
synced with main with rebase.

Will write new tests and better optimize currently working on it

kytpbs committed:
add a test to cancel the next command from the first

Co-authored-by: Joseph Eng <91924258+KangarooKoala@users.noreply.github.com>

KangarooKoala commented on wpilibNewCommands/src/test/native/cpp/frc2/command/CommandScheduleTest.cpp:
```diff
@@ -102,6 +106,43 @@ TEST_F(CommandScheduleTest, NotScheduledCancel) {
+TEST_F(CommandScheduleTest, CancelNextCommand) {
+  CommandScheduler scheduler = GetScheduler();
+
+  frc2::RunCommand** command1PtrPtr = new frc2::RunCommand*;
+  frc2::RunCommand** command2PtrPtr = new frc2::RunCommand*;
+  auto counterPtr = std::make_shared<int>(0);
+
+  auto command1 = frc2::RunCommand([counterPtr, &command2PtrPtr, &scheduler] {
+    scheduler.Cancel(*command2PtrPtr);
+    (*counterPtr)++;
+  });
+  auto command2 = frc2::RunCommand([counterPtr, &command1PtrPtr, &scheduler] {
+    scheduler.Cancel(*command1PtrPtr);
+    (*counterPtr)++;
+  });
+
+  *command1PtrPtr = &command1;
+  *command2PtrPtr = &command2;
```
This can be made clearer by using capture-by-reference in the lambdas. (This also avoids a memory leak from never freeing the pointers allocated by `new`)
```suggestion
  frc2::RunCommand* command1Ptr = nullptr;
  frc2::RunCommand* command2Ptr = nullptr;
  int counter = 0;

  auto command1 = frc2::RunCommand([&counter, &command2Ptr, &scheduler] {
    scheduler.Cancel(command2Ptr);
    counter++;
  });
  auto command2 = frc2::RunCommand([&counter, &command1Ptr, &scheduler] {
    scheduler.Cancel(command1Ptr);
    counter++;
  });

  command1Ptr = &command1;
  command2Ptr = &command2;
```
You'll also need to update the other usages of `counterPtr`.
I think you might find it helpful to think of references as being const pointers (NOT pointers to const) that are automatically dereferenced. Hopefully that explains why you can have the variables store what data it needs (which may be reassigned later) and have the lambdas capture by reference so that when the variable gets a different value, the captured value is also updated.

kytpbs replied:
Thx, finished doing the test at 5 AM where I just wanted to see it work, I will update it to this version with a rebase asap.
For some reason, doing a pass-by-reference didn't work last time, so I did it this way, but let me check again. I may add I'm not used to C++ lambdas so that's why I did that kind of double pointer way. 

Also, IMO the shared pointer makes it easier to debug so I will keep it that way.

kytpbs replied:
added into it with rebase see: 7149e43

kytpbs committed:
add better error messages to C++ tests

kytpbs committed:
Optimize copying `m_scheduledCommands`

kytpbs commented:
fix formatting with rebase + force-push:

kytpbs commented:
Update the checking for null and isScheduled as discussed in discord with a amend + force-push:

kytpbs commented:
oops did `command.isScheduled()` instead of `isScheduled(command)` what a rookie mistake, fixing with force push now

kytpbs committed:
add a test checking if command is scheduled on end() call

kytpbs committed:
Add test for scheduling a command within a command

kytpbs commented:
What other tests should I add? @Starlight220?

KangarooKoala commented on wpilibNewCommands/src/test/java/edu/wpi/first/wpilibj2/command/CommandScheduleTest.java:
```diff
@@ -109,6 +111,102 @@ void schedulerCancelTest() {
+    try (CommandScheduler scheduler = new CommandScheduler()) {
+      Command[] commands = new Command[2];
+      var commandRunCounter = new AtomicInteger(0);
+      // due to the ordering of sets being non-deterministic, we cancel the other command in the
+      // first command and check that the second command is not run
```
This is just me being super nit-picky:
```suggestion
      // for parity with C++ where the ordering of sets is non-deterministic, we cancel the other command in the
      // first command and check that the second command is not run
```
(Formatting probably needs to be changed)

In Java, the ordering of the particular kind of set we use to track commands ([`LinkedHashSet`](https://docs.oracle.com/en/java/javase/17/docs/api/java.base/java/util/LinkedHashSet.html)) is actually deterministic (it's defined to be insertion order). However, it is true that the hashing method used by `SmallPtrSet` in heap allocation mode is non-deterministic (it's based on the pointer's numerical value), so the test should remain written the way it is.

kytpbs replied:
I fully agree with the nitpick, I should have been more specific about it, thx for noticing it

# added in 93ad7b8

KangarooKoala commented on wpilibNewCommands/src/test/java/edu/wpi/first/wpilibj2/command/CommandScheduleTest.java:
```diff
@@ -109,6 +111,102 @@ void schedulerCancelTest() {
+  void scheduleCommandInCommand() {
+    try (CommandScheduler scheduler = new CommandScheduler()) {
+      AtomicInteger counter = new AtomicInteger(0);
+      Command[] commands = new Command[1];
+      Command command1 =
+          new RunCommand(
+              () -> {
+                scheduler.schedule(commands[0]);
+                assertEquals(
+                    1,
+                    counter.get(),
+                    "Command 2's init was not run immediately after getting scheduled");
+              });
+      Command command2 = new InstantCommand(counter::incrementAndGet);
+
+      commands[0] = command2;
```
```suggestion
      Command commandToGetScheduled = new InstantCommand(counter::incrementAndGet);
      Command command =
          new RunCommand(
              () -> {
                scheduler.schedule(commandToGetScheduled);
                assertEquals(
                    1,
                    counter.get(),
                    "Scheduled command's init was not run immediately after getting scheduled");
              });
```
Since we don't have any recursive behavior, we can avoid the array. Also changes the naming to match C++.

kytpbs replied:
My brain focused too much on the previous test, and I did it this way as well. LMAO.

## Added in 93ad7b8

KangarooKoala commented on wpilibNewCommands/src/test/native/cpp/frc2/command/CommandScheduleTest.cpp:
```diff
@@ -95,6 +101,92 @@ TEST_F(CommandScheduleTest, SchedulerCancel) {
+TEST_F(CommandScheduleTest, CancelNextCommand) {
+  CommandScheduler scheduler = GetScheduler();
+
+  frc2::RunCommand* command1PtrPtr = nullptr;
+  frc2::RunCommand* command2PtrPtr = nullptr;
```
```suggestion
  frc2::RunCommand* command1Ptr = nullptr;
  frc2::RunCommand* command2Ptr = nullptr;
```
These are no longer pointer to pointer.

kytpbs replied:
Oops... forgot to change them back after fighting with C++ lambdas

## Added in 93ad7b8

KangarooKoala commented on wpilibNewCommands/src/test/native/cpp/frc2/command/CommandScheduleTest.cpp:
```diff
@@ -95,6 +101,92 @@ TEST_F(CommandScheduleTest, SchedulerCancel) {
+TEST_F(CommandScheduleTest, ScheduleCommandInCommand) {
+  CommandScheduler scheduler = GetScheduler();
+  int counter = 0;
+  frc2::InstantCommand* commandPtr = nullptr;
+
+  auto command = frc2::RunCommand([&] {
+    scheduler.Schedule(commandPtr);
+    EXPECT_EQ(counter, 1)
+        << "Command 2's init was not run immediately after getting scheduled";
+  });
+  auto commandToGetScheduled = frc2::InstantCommand([&counter] { counter++; });
+
+  commandPtr = &commandToGetScheduled;
```
```suggestion
  frc2::InstantCommand commandToGetScheduled{[&counter] { counter++; }};

  auto command = frc2::RunCommand([&] {
    scheduler.Schedule(&commandToGetScheduled);
    EXPECT_EQ(counter, 1)
        << "Scheduled cmmand's init was not run immediately after getting scheduled";
  });
```
See Java. You'll also need to update the usages of `commandPtr` in the `IsScheduled` checks.

kytpbs replied:
## Added in 93ad7b8

KangarooKoala reviewed (commented):
Open for discussion- Should we move some of these tests to `SchedulingRecursionTest`?

kytpbs committed:
Refactor tests according to reviews

Co-Authored-By: Joseph Eng <91924258+KangarooKoala@users.noreply.github.com>

kytpbs commented:
> @KangarooKoala: Should we move some of these tests to SchedulingRecursionTest?

I say we move test `cancelNextCommandTest` for sure and also `scheduleCommandInCommand` since they are both similar to #4259 but keep `commandKnowsWhenEndedTest` since it feels more like a CommandScheduler test more than a scheduling recursion.
Any objections?

KangarooKoala commented on wpilibNewCommands/src/test/java/edu/wpi/first/wpilibj2/command/CommandScheduleTest.java:
```diff
@@ -109,6 +111,102 @@ void schedulerCancelTest() {
+    try (CommandScheduler scheduler = new CommandScheduler()) {
+      Command[] commands = new Command[2];
+      var commandRunCounter = new AtomicInteger(0);
+      // for parity with C++ where the ordering of sets is non-deterministic, we cancel the other
+      // command in the
+      // first command and check that the second command is not run
```
```suggestion
      // for parity with C++ where the ordering of sets is non-deterministic, we cancel the other
      // command in the first command and check that the second command is not run
```
Makes the comment easier to read.

kytpbs replied:
added in ada142a

KangarooKoala reviewed (commented):
> I say we move test cancelNextCommandTest for sure and also scheduleCommandInCommand since they are both similar to https://github.com/wpilibsuite/allwpilib/issues/4259 but keep commandKnowsWhenEndedTest since it feels more like a CommandScheduler test more than a scheduling recursion.
Any objections?

Sounds good!

KangarooKoala reviewed (commented):
Something weird happened with my internet while I was submitting my review, guess it submitted twice.

kytpbs committed:
move some tests to `schedulingRecursionTest` & improve comments

kytpbs commented:
Force pushing to rebase with main...

kytpbs commented:
Oop, left some errors while rebasing, force-pushing again to fix those ASAP

kytpbs commented:
Unintentionally removed `ownedCommands`, adding it back with rebase now

rzblue commented on wpilibNewCommands/src/main/java/edu/wpi/first/wpilibj2/command/CommandScheduler.java:
```diff
@@ -274,11 +264,19 @@ public void run() {
+    m_scheduledCommandsCopy = m_scheduledCommands.toArray(m_scheduledCommandsCopy);
+    for (Command command : m_scheduledCommandsCopy) {
+      if (command == null) {
+        // at least 1 Command was removed before `run` was called (diff between copy and original).
```
```suggestion
```
This isn't the case- null should only be present in the array after the last element of the set

kytpbs replied:
Oops, it was stuck there from the previous implementation, thanks for pointing it out fixing it now.

rzblue commented on wpilibNewCommands/src/main/native/cpp/frc2/command/CommandScheduler.cpp:
```diff
@@ -224,36 +214,22 @@ void CommandScheduler::Run() {
       }
 
       m_watchdog.AddEpoch(command->GetName() + ".End(false)");
+
```
```suggestion
```

rzblue commented on wpilibNewCommands/src/test/native/cpp/frc2/command/CommandScheduleTest.cpp:
```diff
@@ -2,6 +2,12 @@
 // Open Source Software; you can modify and/or share it under the terms of
 // the WPILib BSD license file in the root directory of this project.
 
+#include <frc2/command/FunctionalCommand.h>
+#include <frc2/command/InstantCommand.h>
+#include <frc2/command/RunCommand.h>
```
use quotes instead of angle brackets

rzblue commented on wpilibNewCommands/src/test/native/cpp/frc2/command/CommandScheduleTest.cpp:
```diff
@@ -2,6 +2,12 @@
+#include <frc2/command/InstantCommand.h>
+#include <frc2/command/RunCommand.h>
+
+#include <memory>
```
not needed

kytpbs committed:
Remove Unnecessary parts + correct imports

Co-Authored-By: Ryan Blue <13878527+rzblue@users.noreply.github.com>

KangarooKoala commented:
Now that the season's done, could we get a review of this?

Starlight220 commented on wpilibNewCommands/src/main/java/edu/wpi/first/wpilibj2/command/CommandScheduler.java:
```diff
@@ -291,32 +288,17 @@ public void run() {
       m_watchdog.addEpoch(command.getName() + ".execute()");
       if (command.isFinished()) {
-        m_endingCommands.add(command);
+        // remove command first so the command calling cancel() doesn't crash...
```
Why would `cancel()` crash?

Currently `isScheduled` is defined to return true until after `end` is called; this changes that behavior and can cause some issues (see the SchedulingRecursion test suite; I can't find the gh issues at the moment).

kytpbs replied:
I believed that this behavior change represents more of `isScheduled` since when a command is ending, it is also getting unscheduled, meaning `isScheduled` returning false makes more sense.

kytpbs replied:
> Looks good to me! (For what it's worth) I'll just note down the behavioral differences for posterity, but I think these are acceptable (even positive) changes:
> 
> * `command.isScheduled()` is `false` when `command.end(boolean)` and the finish or interrupt callbacks are called.
> * When scheduling a command from the run loop, `command.initialize()` and the schedule callbacks are called immediately, not after the run loop. It remains the case that `command.execute()` will only be called from the next call to `scheduler.run()`.
> * When cancelling a command from the run loop , `command.end(true)` and the finish or interrupt callbacks are called immediately, not after the run loop (and the queued scheduled commands are processed). Notably, if a command (call it A) cancels another command (call it B) that was scheduled after A (and therefore will be processed after A), B is **not** be processed (`execute()` called and `isFinished()` checked) in the same run loop.
https://github.com/wpilibsuite/allwpilib/pull/6593#pullrequestreview-2069797402

This behaviour change and more are also mentioned in this comment

KangarooKoala replied:
If the command is still scheduled when `end()` is called, then if it cancels itself (potentially transitively), `end()` will be called again, leading to at least a violation of the lifecycle conditions and probably boundless recursion (causing the crash that was mentioned).

We should definitely make the comment more specific.

Starlight220 commented on wpilibNewCommands/src/main/java/edu/wpi/first/wpilibj2/command/CommandScheduler.java:
```diff
@@ -467,25 +449,16 @@ private void cancel(Command command, Optional<Command> interruptor) {
     }
 
-    m_endingCommands.add(command);
+    // remove command first so the command calling cancel() on itself doesn't crash...
```
Ditto

Starlight220 commented on wpilibNewCommands/src/main/java/edu/wpi/first/wpilibj2/command/CommandScheduler.java:
```diff
@@ -274,11 +264,18 @@ public void run() {
+    m_scheduledCommandsCopy = m_scheduledCommands.toArray(m_scheduledCommandsCopy);
+    for (Command command : m_scheduledCommandsCopy) {
+      if (command == null) {
+        // No more elements to iterate over (see toArray documentation)
```
I'm not sure that basing functionality on some side effect of a function is a good idea.


Why not use a proper Collection rather than an array?

kytpbs replied:
Because addAll under the hood uses `c.toArray()`, which means it will allocate a new array every time it's called, independent from the size of `m_scheduledCommandsCopy`.

So if we are gonna do that, we may as well just do `m_scheduledCommands.toArray()` in the for loop and not have the `m_scheduledCommandsCopy` at all, but I feel like it would be a big performance bottleneck

Source:
https://github.com/openjdk/jdk/blob/f6c90fe8f9986b02797ba1f967c71a592a892266/src/java.base/share/classes/java/util/ArrayList.java#L752

Gold856 committed:
Merge branch 'main' into Fix-cpp-commandScheduler

Gold856 committed:
Copy Python port

Gold856 committed:
Merge branch 'main' into Fix-cpp-commandScheduler

Gold856 committed:
Merge branch 'main' into Fix-cpp-commandScheduler

kytpbs commented:
@Gold856 

For the python port that was added into the branch should we do a similar implementation to
```java
private Command[] m_scheduledCommandsCopy = new Command[12];
...
m_scheduledCommandsCopy = m_scheduledCommands.toArray(m_scheduledCommandsCopy);
```

with
```python
def init(...):
    ...
    self._scheduled_copy = []
    # Possibly = [None] * 12; .empty() to pre-allocate similar to Java

...

self._scheduled_copy[:] = self._scheduled
```

In GIL CPython when you copy things over this way, it works similar to .toArray in the sense that it will change the size of the list and realloc only if the size of the array increases, which means it not allocate as long as scheduled commands is less then the array size (it does do a unalloc if the scheduled_copy is >2x bigger then scheduled, which is different from Java setting to null and leaving it be)

So we could use that if we wish to make things closer to Java in terms of behavior and source code similarity

HOWEVER, tho considering the size .copy uses memory recovered from the free list inside the interpreter so not much of a performance difference and more of a semantics difference...

.copy could possibly be more optimized since slice copying needs to run more code in CPython's side.

kytpbs commented:
Also according to Java's .toArray implementation it will not put null from `scheduledCommands.size()` to `scheduledCommandsCopy.size()` but instead will only put null once at index: `scheduledCommands.size()`.

This means we will hold stale references to commands inside, which differs from the Cpp implementation and could be considered a memory leak if:

`commandScheduler.schedule(new SomeCommand())` gets called 

Because now we will forever hold on to SomeCommand and it will never get GC'd, Even if this is the unrecommended way to use the command scheduler, this is possibly a bug

Diagram:

```java
scheduledCommands.size() == 3; scheduledCommandsCopy.size() == 12;
[0] cmd0
[1] cmd1
[2] cmd2
[3] null      <- the guaranteed null
[4] unscheduledCmd0 <- still strongly reachable, GC can't collect
...
[11] unscheduledCmd7 <- strong reference, GC can't collect
```
